src/types/acls.rs

```diff
@@ -0,0 +1,350 @@
+        }
+
+        Ok(input
+            .chars()
+            .map(|c| c.into())
+            .collect::<HashSet<AclRight>>()
+            .into())
+    }
+}
+
+/// Error from parsing AclRight strings
+#[derive(Debug, Eq, PartialEq)]
+#[non_exhaustive]
+pub enum AclRightError {
+    /// Returned when a non-lower-case alpha numeric is provided in the rights list string.
+    InvalidRight,
+}
+
+impl Display for AclRightError {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match *self {
+            AclRightError::InvalidRight => {
+                write!(f, "Rights may only be lowercase alpha numeric characters")
+            }
+        }
+    }
+}
+
+impl std::error::Error for AclRightError {}
+
+/// From [section 3.6 of RFC 4313](https://datatracker.ietf.org/doc/html/rfc4314#section-3.6).
+///
+/// The ACL response from the [`Session::get_acl`] IMAP command
+#[self_referencing]
+pub struct Acl {
+    data: Vec<u8>,
+    #[borrows(data)]
+    #[covariant]
+    pub(crate) acl: InnerAcl<'this>,
+}
+
+impl Acl {
+    /// Parse the given input into a [`Acl`] response.
+    pub fn parse(
+        owned: Vec<u8>,
+        unsolicited: &mut mpsc::Sender<UnsolicitedResponse>,
+    ) -> Result<Self, Error> {
+        AclTryBuilder {
+            data: owned,
+            acl_builder: |input| {
+                // There should only be ONE single ACL response
+                parse_until_done(input, unsolicited, |response| match response {
+                    Response::Acl(a) => Ok(MapOrNot::Map(InnerAcl {
+                        mailbox: a.mailbox,
+                        acls: a
+                            .acls
+                            .into_iter()
+                            .map(|e| AclEntry {
+                                identifier: e.identifier,
+                                rights: e.rights.into(),
+                            })
+                            .collect(),
+                    })),
+                    resp => Ok(MapOrNot::Not(resp)),
+                })
+            },
+        }
+        .try_build()
+    }
+
+    /// Return the mailbox the ACL entries belong to
+    pub fn mailbox(&self) -> &str {
+        &*self.borrow_acl().mailbox
+    }
+
+    /// Returns a list of identifier/rights pairs for the mailbox
+    pub fn acls(&self) -> &[AclEntry<'_>] {
+        &*self.borrow_acl().acls
+    }
+}
+
+/// Inner struct to manage storing the references for ouroboros
+#[derive(Debug, Eq, PartialEq)]
+pub(crate) struct InnerAcl<'a> {
+    /// The mailbox the ACL Entries belong to
+    pub(crate) mailbox: Cow<'a, str>,
+    /// The list of identifier/rights pairs for the mailbox
+    pub(crate) acls: Vec<AclEntry<'a>>,
+}
+
+/// From [section 3.6 of RFC 4313](https://datatracker.ietf.org/doc/html/rfc4314#section-3.6).
+///
+/// The list of identifiers and rights for the [`Acl`] response
+#[derive(Debug, Eq, PartialEq, Clone)]
+pub struct AclEntry<'a> {
```

Comment:
Let's set this to be `#[non_exhaustive]` in case other extensions we add support for in the future might add more fields.

Reply:
This still needs to be done.

Reply:
I added it to the wrong structs :)